python/cog/server/http.py

```diff
@@ -111,6 +127,86 @@ def healthcheck() -> Any:
             }
         )
 
+    @app.post(
+        "/trainings",
+        response_model=TrainingResponse,
+        response_model_exclude_unset=True,
+    )
+    def train(request: TrainingRequest = Body(default=None), prefer: Union[str, None] = Header(default=None)) -> Any:  # type: ignore
+        """
+        Run a training on the model
+        """
+        if mode != "train":
+            return JSONResponse(
+                {"detail": "The application is running as a predictor, not a trainer."},
+                status_code=400,
+            )
+        if runner.is_busy():
+            return JSONResponse({"detail": "Already training."}, status_code=409)
+
+        respond_async = prefer == "respond-async"
+
+        return _train(request=request, respond_async=respond_async)
+
+    def _train(*, request: TrainingRequest, respond_async: bool) -> Response:
+        # [compat] If no body is supplied, assume that this model can be run
+        # with empty input. This will throw a ValidationError if that's not
+        # possible.
+        if request is None:
+            request = TrainingRequest(input={})
+        # [compat] If body is supplied but input is None, set it to an empty
+        # dictionary so that later code can be simpler.
+        if request.input is None:
+            request.input = {}
+
+        try:
+            # For now, we only ask the Runner to handle file uploads for
+            # async predictions. This is unfortunate but required to ensure
+            # backwards-compatible behaviour for synchronous predictions.
+            initial_response, async_result = runner.run(request, upload=respond_async)
+        except RunnerBusyError:
+            return JSONResponse(
+                {"detail": "Already running training."}, status_code=409
+            )
+
+        if respond_async:
+            return JSONResponse(jsonable_encoder(initial_response), status_code=202)
+
+        try:
+            response = TrainingResponse(**async_result.get().dict())
+        except ValidationError as e:
+            _log_invalid_output(e)
+            raise HTTPException(status_code=500)
+
+        response_object = response.dict()
+        response_object["output"] = upload_files(
+            response_object["output"],
+            upload_file=lambda fh: upload_file(fh, request.output_file_prefix),  # type: ignore
```

Comment:
Similarly, we can probably get rid of this part given that we don't support `output_file_prefix` on `TrainingRequest`s if I read the earlier code correctly. That would then mean that we can pass in `upload=True` to `runner.run(...)` a few lines earlier.

Reply:
Removed, and set `upload=True`.